packages/cdktf-cli/lib/dependencies/dependency-manager.ts

```diff
@@ -0,0 +1,260 @@
+import { Language } from "@cdktf/provider-generator";
+import { toPascalCase } from "codemaker";
+import { ProviderDependencySpec } from "../cdktf-config";
+import { Errors } from "../errors";
+import { logger } from "../logging";
+import { CdktfConfigManager } from "./cdktf-config-manager";
+import { PackageManager } from "./package-manager";
+import {
+  getNpmPackageName,
+  getPrebuiltProviderVersion,
+} from "./prebuilt-providers";
+import { getLatestVersion } from "./registry-api";
+import { versionMatchesConstraint } from "./version-constraints";
+
+// ref: https://www.terraform.io/language/providers/requirements#source-addresses
+const DEFAULT_HOSTNAME = "registry.terraform.io";
+const DEFAULT_NAMESPACE = "hashicorp";
+function normalizeProviderSource(source: string) {
```

Comment:
Isn't this quite similar? https://github.com/hashicorp/terraform-cdk/blob/main/packages/%40cdktf/provider-generator/lib/config.ts#L75

Reply:
Similar as in "we should refactor this into a single place" yes, but the one you linked also parses versions etc. Also: the add command currently does not support local providers or providers from VCS, so that's something for the future as well :)

Reply:
Oh I didn't know, we need to make sure we document this and possible detect it and throw errors.